### PR TITLE
 fix(wayland): register frame callback then commit (fixes #3468) 

### LINKED ISSUE #3468: wayland: Pressing and holding a button is laggy when using the webgpu frontend
### What Operating System(s) are you seeing this problem on?

Linux Wayland

### Which Wayland compositor or X11 Window manager(s) are you using?

Mutter

### WezTerm version

20230407-221507-36d5307b

### Did you try the latest nightly build to see if the issue is better (or worse!) than your current version?

Yes, and I updated the version box above to show the version of the nightly that I tried

### Describe the bug

After the latest commit, the webgpu frontend has become usable on Linux.
But pressing and holding a key is really laggy now.
Whether its holding backspace to delete, a character to type, or even arrow keys to navigate.

This only happens in the wayland and not in the xwayland version 

### To Reproduce

Open wezterm and press and hold any key

### Configuration

local wezterm = require 'wezterm';
local act = wezterm.action

return {
  window_decorations = "INTEGRATED_BUTTONS|RESIZE",
  integrated_title_button_style = "Gnome",
  color_scheme = "Kasugano (terminal.sexy)",
  enable_wayland = true,
  hide_mouse_cursor_when_typing = false,
  front_end = 'WebGpu'
}

### Expected Behavior

The action happens continuously and smoothly instead of being laggy.

### Logs

```
wezterm version: 20230407-221507-36d5307b x86_64-unknown-linux-gnu
Window Environment: Wayland
OpenGL version: WebGPU: name=AMD Radeon Graphics (RADV RENOIR), device_type=IntegratedGpu, backend=Vulkan, driver=radv, driver_info=Mesa 23.0.1, vendor=4098, device=5688
```
Launching with `WAYLAND_DEBUG=1 wezterm` and holding a key gives the following loop:

```
[ 659559.337] wl_buffer@48.release()
[ 659574.863] wl_display@1.delete_id(78)
[ 659574.885] wl_callback@78.done(5654499)
[ 659574.899]  -> wl_surface@20.attach(wl_buffer@48, 0, 0)
[ 659574.918]  -> wl_surface@20.damage(0, 0, 2147483647, 2147483647)
[ 659574.931]  -> wl_surface@20.frame(new id wl_callback@78)
[ 659574.943]  -> wl_surface@20.commit()
[ 659575.768] wl_buffer@69.release()
[ 659591.372] wl_display@1.delete_id(78)
[ 659591.400] wl_callback@78.done(5654516)
[ 659591.420]  -> wl_surface@20.attach(wl_buffer@69, 0, 0)
[ 659591.433]  -> wl_surface@20.damage(0, 0, 2147483647, 2147483647)
[ 659591.448]  -> wl_surface@20.frame(new id wl_callback@78)
[ 659591.459]  -> wl_surface@20.commit()
[ 659592.027]  -> zwp_text_input_v3@14.set_cursor_rectangle(640, 63, 10, 20)
[ 659592.042]  -> zwp_text_input_v3@14.commit()
[ 659592.563] wl_buffer@48.release()
[ 659607.587] wl_display@1.delete_id(78)
[ 659607.611] wl_callback@78.done(5654532)
[ 659607.630]  -> wl_surface@20.attach(wl_buffer@48, 0, 0)
[ 659607.644]  -> wl_surface@20.damage(0, 0, 2147483647, 2147483647)
[ 659607.652]  -> wl_surface@20.frame(new id wl_callback@78)
[ 659607.664]  -> wl_surface@20.commit()
[ 659608.452] zwp_text_input_v3@14.done(46)
[ 659609.699] wl_buffer@69.release()
[ 659624.235] wl_display@1.delete_id(78)
[ 659624.251] wl_callback@78.done(5654549)
[ 659624.258]  -> wl_surface@20.attach(wl_buffer@69, 0, 0)
[ 659624.269]  -> wl_surface@20.damage(0, 0, 2147483647, 2147483647)
[ 659624.275]  -> wl_surface@20.frame(new id wl_callback@78)
[ 659624.279]  -> wl_surface@20.commit()
[ 659624.849] wl_buffer@48.release()
[ 659639.689] wl_display@1.delete_id(78)
[ 659639.698] wl_callback@78.done(5654564)
[ 659639.702]  -> wl_surface@20.attach(wl_buffer@48, 0, 0)
[ 659639.705]  -> wl_surface@20.damage(0, 0, 2147483647, 2147483647)
[ 659639.708]  -> wl_surface@20.frame(new id wl_callback@78)
[ 659639.710]  -> wl_surface@20.commit()
[ 659640.055] wl_buffer@69.release()
[ 659658.082] wl_display@1.delete_id(78)
[ 659658.112] wl_callback@78.done(5654582)
[ 659658.123]  -> wl_surface@20.attach(wl_buffer@69, 0, 0)
[ 659658.132]  -> wl_surface@20.damage(0, 0, 2147483647, 2147483647)
[ 659658.139]  -> wl_surface@20.frame(new id wl_callback@78)
[ 659658.145]  -> wl_surface@20.commit()
[ 659659.266] wl_buffer@48.release()
[ 659673.854] wl_display@1.delete_id(78)
[ 659673.879] wl_callback@78.done(5654598)
[ 659673.900]  -> wl_surface@20.attach(wl_buffer@48, 0, 0)
[ 659673.914]  -> wl_surface@20.damage(0, 0, 2147483647, 2147483647)
[ 659673.924]  -> wl_surface@20.frame(new id wl_callback@78)
[ 659673.936]  -> wl_surface@20.commit()
[ 659674.837] wl_buffer@69.release()
[ 659690.713] wl_display@1.delete_id(78)
[ 659690.721] wl_callback@78.done(5654615)
[ 659690.725]  -> wl_surface@20.attach(wl_buffer@69, 0, 0)
[ 659690.729]  -> wl_surface@20.damage(0, 0, 2147483647, 2147483647)
[ 659690.733]  -> wl_surface@20.frame(new id wl_callback@78)
[ 659690.735]  -> wl_surface@20.commit()
[ 659691.053] wl_buffer@48.release()
[ 659707.664] wl_display@1.delete_id(78)
[ 659707.689] wl_callback@78.done(5654632)
[ 659707.700]  -> wl_surface@20.attach(wl_buffer@48, 0, 0)
[ 659707.709]  -> wl_surface@20.damage(0, 0, 2147483647, 2147483647)
[ 659707.717]  -> wl_surface@20.frame(new id wl_callback@78)
[ 659707.723]  -> wl_surface@20.commit()
[ 659708.797] wl_buffer@69.release()
[ 659724.595] wl_display@1.delete_id(78)
[ 659724.618] wl_callback@78.done(5654649)
[ 659724.636]  -> wl_surface@20.attach(wl_buffer@69, 0, 0)
[ 659724.650]  -> wl_surface@20.damage(0, 0, 2147483647, 2147483647)
[ 659724.660]  -> wl_surface@20.frame(new id wl_callback@78)
[ 659724.671]  -> wl_surface@20.commit()
[ 659725.363]  -> zwp_text_input_v3@14.set_cursor_rectangle(650, 63, 10, 20)
[ 659725.382]  -> zwp_text_input_v3@14.commit()
````

### Anything else?

I don't have any knowledge about this, but for it to be smooth,
```
[ 659725.363]  -> zwp_text_input_v3@14.set_cursor_rectangle(650, 63, 10, 20)
[ 659725.382]  -> zwp_text_input_v3@14.commit()
```
Should be present in every loop?

### PR DESCRIPTION
**Pull request by jokeyrhyme**
Fixes #3468

### CODE REVIEW

File: window/src/os/wayland/window.rs [hunk 1]
```diff
@@ -801,27 +801,22 @@ impl WaylandWindowInner {
...
+                inner.next_frame_is_ready();
+                Ok(())
             });
-            self.frame_callback.replace(callback);
-        }
+        });
+        self.frame_callback.replace(callback);
+
+        // commit pending surface state,
+        // ensuring the compositor eventually wakes us up via the above callback
```
Comment (wez) [inline window/src/os/wayland/window.rs:818]:
```suggestion
        // The repaint has the side of effect of committing the surface,
        // which is necessary for the frame callback to get triggered.
        // Ordering the repaint after requesting the callback ensures that
        // we will get woken at the appropriate time.
        // <https://github.com/wez/wezterm/issues/3468>
        // <https://github.com/wez/wezterm/issues/3126>
```

**Review by wez — CHANGES_REQUESTED** [review]
Thanks for this!

The changelog appears to have been reformatted substantially. Please revert those changes and keep this PR focused on the rust file change.

File: window/src/os/wayland/window.rs (hunk 1 above)
Comment (jokeyrhyme) [reply window/src/os/wayland/window.rs:818]: Ah, sorry about that, the revert had a conflict in docs/changelog.md and my editor auto-formatted it when I fixed the conflict

**Comment by jokeyrhyme** [comment]
I've just squashed, the other thing I noticed is the freebsd-13 build keeps failing, I've retried it a few times now :shrug:

**Comment by VarLad** [comment]
@jokeyrhyme I believe FreeBSD 13 breaking is not your PR since you can check, its been breaking for a lot of the past commits as well. 
Like here, in the latest commit as of writing https://github.com/wez/wezterm/commit/54a65a5401a86bab5c56751abca7de438c408b4a

**Pull request merged**